paddle/fluid/framework/ir/graph_pattern_detector.h

```diff
@@ -523,6 +523,45 @@ struct ElewiseAddActInplaceGrad : public PatternBase {
   PATTERN_DECL_NODE(d_ele_y);
   PATTERN_DECL_NODE(ele_y);
 };
+
+// Convolution op
+// Forward pass for convolution.
+// conv_input, conv_bias and conv_filter are inputs.
+// conv_output is a result of the operator.
+// residual_data is data used by skip connection.
+// If residual connection fusion is on, the formula is:
+// conv_output = conv_op(conv_filter, conv_input, conv_bias)
+//             + conv_residual_data
+// If the fusion is off, conv_residual_data is not added.
+struct Conv : public PatternBase {
+  Conv(PDPattern* pattern, const std::string& name_scope)
+      : PatternBase(pattern, name_scope, "convolution") {}
+
+  PDNode* operator()();
```

Comment:
I corrected it: I added `AsInput` or `AsOutput` method calls when new nodes in a pattern are created.

I would like to point out that `AsInput` and `AsOutput` set member field `role_`, that is used by `IsInput` and `IsOutput` methods. These methods are not used in graph_pattern_detector code, nor is `role_` field.

Another thing is that a node can be used both as input and as output depending on the role it is used in a pattern: convolution output is marked as output in a convolution pattern, but it could be marked as input in an elementwise_add pattern.

Reply:
Thank you for explanaition!